reposerver/repository/repository.go

```diff
@@ -163,6 +163,11 @@ func (s *Service) runRepoOperation(
 		if err != nil {
 			return nil, err
 		}
+		sha, err := gitClient.CommitSHA()
```

Comment:
I am skeptical to this fix. `gitClient.CommitSHA()` gives us the SHA of the HEAD which may not be the resolving tag is pointing to.   I think we want to figure out what caused  SHA resolution hehavior change since v1.4.3.

Reply:
I just tested it again to be sure, it is picking up the correct commit SHA for the specified tag even when HEAD points to a different SHA 
maybe it's using the state of the commit history from a particular revision so HEAD resolves to the same as that commit?

As for the behaviour change, from what I understand, the ```lsRemote``` function just runs ```git ls-remote```'s output, which is the list of all refs and tags with their commits. In this list, it seems the annotated tag's hash always appears before the SHA of the commit it points to (which is in the next line below it) and the current code just traverses the list and picks up the first one it sees, and ends up taking the hash of the tag instead of the commit SHA

if I had to guess, maybe the commit SHA showed before the tag hash at some point and the ordering of the hashes present in the output of ls-remote may have changed recently (since the argocd code for this doesn't seem to have changed between 1.4.2 and the current version)

Reply:
IMHO, the only way to get the actual revision of the commit an annotated tag is pointing to is to checkout the tag, bringing the repository into _detached HEAD_ state, and then resolve HEAD revision. 

The checkout to the target revision is done a little later down the code path (https://github.com/argoproj/argo-cd/blob/3ae1d13dfd541c34a2b970b7a6d8f32f2369d31a/reposerver/repository/repository.go#L213). So probably the current incarnation of this PR will only work if the annotated tag is actually pointing to the real HEAD of the repository's default branch.

Reply:
@jaanfis 
The fix is working even when HEAD points to a different commit SHA
Seeing some strange behavior I haven't been able to explain yet though, ```CommitSHA()``` seems to be running ```git rev-parse HEAD``` but returning the correct commit SHA almost as if the repo is somehow already being checked out even before the actual call to ```checkoutRevision``` happens 
Even when immediately before/after HEAD does get resolved to the actual HEAD in a different function call 
There seems to be some kind of inconsistency between the api-server and repo-server here as to what HEAD actually points to
```
22:09:33  api-server | DEBU[0040] symbolic reference 'HEAD' (refs/heads/master) resolved to '401ea4c66bbd95c18eb3ed4d05912514cae83f30' 
22:09:33 repo-server | DEBU[0040]                                               duration=1.082374781s execID=OAWIY
22:09:33 repo-server | INFO[0040] git checkout --force 95249be61b028d566c29d47b19e65c5603388a41  dir="/tmp/https:__github.com_jaideepr97_argocd-experiment" execID=GH1HK
22:09:33 repo-server | DEBU[0040]                                               duration=3.359517ms execID=GH1HK
22:09:33 repo-server | INFO[0040] git clean -fdx                                dir="/tmp/https:__github.com_jaideepr97_argocd-experiment" execID=usYfT
22:09:33 repo-server | DEBU[0040]                                               duration=2.442886ms execID=usYfT
22:09:33 repo-server | INFO[0040] git rev-parse HEAD                            dir="/tmp/https:__github.com_jaideepr97_argocd-experiment" execID=J2b1A
22:09:33 repo-server | DEBU[0040] 632039659e542ed7de0c170a4fcc1c571b288fc0      duration=3.721488ms execID=J2b1A
```

Reply:
@jaanfis
The fix is working even when HEAD points to a different commit SHA
Seeing some strange behavior I haven't been able to explain yet though, ```CommitSHA()``` seems to be running ```git rev-parse HEAD``` but returning the correct commit SHA almost as if the repo is somehow already being checked out even before the actual call to ```checkoutRevision``` happens, which happens later like you pointed out
Even when immediately before/after HEAD does get resolved to the actual HEAD in a different function call
There seems to be some kind of inconsistency between the api-server and repo-server here as to what HEAD actually points to
```
22:09:32 repo-server | DEBU[0039] revision 'v0.0.8' resolved to '95249be61b028d566c29d47b19e65c5603388a41' 
22:09:32 repo-server | INFO[0039] git fetch origin --tags --force               dir="/tmp/https:__github.com_jaideepr97_argocd-experiment" execID=OAWIY
22:09:32 repo-server | DEBU[0039] revision 'v0.0.8' resolved to '95249be61b028d566c29d47b19e65c5603388a41' 
22:09:32 repo-server | INFO[0039] git rev-parse HEAD                            dir="/tmp/https:__github.com_jaideepr97_argocd-experiment" execID=W212M
22:09:32 repo-server | DEBU[0039] 632039659e542ed7de0c170a4fcc1c571b288fc0      duration=3.83898ms execID=W212M
22:09:32 repo-server | INFO[0039] app details cache miss: 632039659e542ed7de0c170a4fcc1c571b288fc0/&ApplicationSource{RepoURL:https://github.com/jaideepr97/ARGOCD-EXPERIMENT,Path:helm-guestbook,TargetRevision:v0.0.8,Helm:nil,Kustomize:nil,Ksonnet:nil,Directory:nil,Plugin:nil,Chart:,} 
22:09:32  controller | INFO[0040] Successfully saved info of 1 clusters        
22:09:32  api-server | 2020/10/15 22:09:32 proto: tag has too few fields: "-"
22:09:32  api-server | INFO[0039] received unary call /application.ApplicationService/Create  grpc.method=Create grpc.request.claims=null grpc.request.content="%!v(PANIC=String method: reflect.Value.Bytes of non-byte slice)" grpc.service=application.ApplicationService grpc.start_time="2020-10-15T22:09:32Z" span.kind=server system=grpc
22:09:33  api-server | DEBU[0040] symbolic reference 'HEAD' (refs/heads/master) resolved to '401ea4c66bbd95c18eb3ed4d05912514cae83f30'
```

Here  ```401ea4c66bbd95c18eb3ed4d05912514cae83f30``` is the actual HEAD, ```95249be61b028d566c29d47b19e65c5603388a41``` is the hash of the annotated tag and ```632039659e542ed7de0c170a4fcc1c571b288fc0``` is the commit I want

Reply:
@jaideepr97
It is a regression introduced in 1.6.0.  
The revision returned from checkoutRevision() was ignored.  This could fix it.
https://github.com/wtam2018/argo-cd/commit/3b6135418affecdc5a8f6ec715ad19e41494e81b

Reply:
@alexmt Do you recall the reason to ignore the revision returned from `checkoutRevision()`?
https://github.com/argoproj/argo-cd/commit/4f49168c1a61e662c90b2a24c8cdbc023c8273d2

Reply:
Thank you for finding the root cause @wtam2018 , @jaideepr97 . Finally found why changes https://github.com/argoproj/argo-cd/commit/4f49168c1a61e662c90b2a24c8cdbc023c8273d2 was implemented. 
That was an attempt to fix caching for git tags. The problem is that we were using tag hash, returned by `ls-remote` to load cached manifests:

https://github.com/argoproj/argo-cd/blob/15b0a6e793d8f2c2ac62617135603081938a79ba/reposerver/repository/repository.go#L169

and then commit sha to store generated manifests in cache:

https://github.com/argoproj/argo-cd/blob/15b0a6e793d8f2c2ac62617135603081938a79ba/reposerver/repository/repository.go#L227

(`operation` eventually uses `revision` to store cache). 

So repo-server was unable to property cache manifests for annotated tags and had to regenerate them again and again.

I think this PR re-introduced old bug again. @jaideepr97 I think we can fix it by adding one more parameter to `operation` that holds `ls-remote` result, so that `operation` has both commitSHA and revision used to retrieve cache.